[UnifiedPDF] [iOS] Inflate the plugin to fit the PDF content

aprotyas opened the pull request:
#### 2c1a883b5e251c7e4f932ebbd9280485b7b3696c
<pre>
[UnifiedPDF] [iOS] Inflate the plugin to fit the PDF content
<a href="https://bugs.webkit.org/show_bug.cgi?id=285765">https://bugs.webkit.org/show_bug.cgi?id=285765</a>
<a href="https://rdar.apple.com/142516290">rdar://142516290</a>

Reviewed by Wenson Hsieh.

For iOS, we want to use main frame scrolling rather than have the
plugin as a subscrollable region. This means that the plugin needs to be
inflated to fit the PDF content size.

This patch introduces the first step in this direction. Future patches
will involve, among other things, changes to how the plugin handles page
scale gestures, to maintain the &quot;plugin fits content&quot; behavior on scale
gestures or programmatic view scale updates.

* Source/WebCore/html/PluginDocument.cpp:
(WebCore::PluginDocumentParser::createStyleElement):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::shouldSizeToFitContent const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::installPDFDocument):
(WebKit::UnifiedPDFPlugin::shouldSizeToFitContent const):
(WebKit::UnifiedPDFPlugin::sizeToFitContentsIfNeeded):
(WebKit::UnifiedPDFPlugin::updateLayout):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::PluginView):
(WebKit::PluginView::updateDocumentForPluginSizingBehavior):

Canonical link: <a href="https://commits.webkit.org/288932@main">https://commits.webkit.org/288932@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/68b71d72f964ee8ff869174ff645271e8e034596

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84811 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4536 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39199 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89950 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35863 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86896 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4625 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12511 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65993 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23823 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87856 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3513 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77074 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46268 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3392 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31296 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34937 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32105 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91326 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12150 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8867 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74476 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12377 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72888 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73601 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18212 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17982 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16431 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/3598 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12102 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/17564 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11936 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15430 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13682 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -333,6 +333,28 @@ static String mutationObserverNotificationString()
+
+    auto& pluginElement = m_view->pluginElement();
+    pluginElement.setInlineStyleProperty(CSSPropertyWidth, size.width(), CSSUnitType::CSS_PX);
+    pluginElement.setInlineStyleProperty(CSSPropertyHeight, size.height(), CSSUnitType::CSS_PX);
```
```suggestion
    Ref pluginElement = m_view->pluginElement();
    pluginElement->setInlineStyleProperty(CSSPropertyWidth, size.width(), CSSUnitType::CSS_PX);
    pluginElement->setInlineStyleProperty(CSSPropertyHeight, size.height(), CSSUnitType::CSS_PX);
```

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/e6abaffdcea753bb1711d92bc067f42526af4c96

pxlcoder commented on Source/WebCore/html/PluginDocument.cpp:
```diff
@@ -76,7 +76,11 @@ Ref<HTMLStyleElement> PluginDocumentParser::createStyleElement(Document& documen
+html, body, embed { width: 100%; height: 100%; }
+body { margin: 0; overflow: hidden; }
+html.plugin-fits-content body { overflow: revert; }
+)CONTENTS"_s;
```
Can we indent this?

aprotyas replied:
Yes, I will

aprotyas replied:
Indented in https://github.com/WebKit/WebKit/commit/e6abaffdcea753bb1711d92bc067f42526af4c96

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -215,6 +215,8 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
     double minScaleFactor() const final;
     double maxScaleFactor() const final;
 
+    bool pluginFitsContent() const final;
```
Should we call this `shouldSizeToFitContent`?

aprotyas replied:
Sure!

aprotyas replied:
Changed in https://github.com/WebKit/WebKit/commit/e6abaffdcea753bb1711d92bc067f42526af4c96

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -333,6 +333,28 @@ static String mutationObserverNotificationString()
     if (m_pdfTestCallback)
         std::exchange(m_pdfTestCallback, nullptr)->handleEvent();
+
+    sizeToFitContents();
```
Is it ok to do this after we run the tests or should we do them before?

aprotyas replied:
That's a good point. In fact, the test callback should be the last thing to run once we are certain that the document is installed and in a valid state.

aprotyas replied:
I'll make that bit a scope exit.

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/e6abaffdcea753bb1711d92bc067f42526af4c96

aprotyas committed:
[UnifiedPDF] [iOS] Inflate the plugin to fit the PDF content
https://bugs.webkit.org/show_bug.cgi?id=285765
rdar://142516290

Reviewed by Wenson Hsieh.

For iOS, we want to use main frame scrolling rather than have the
plugin as a subscrollable region. This means that the plugin needs to be
inflated to fit the PDF content size.

This patch introduces the first step in this direction. Future patches
will involve, among other things, changes to how the plugin handles page
scale gestures, to maintain the "plugin fits content" behavior on scale
gestures or programmatic view scale updates.

* Source/WebCore/html/PluginDocument.cpp:
(WebCore::PluginDocumentParser::createStyleElement):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::shouldSizeToFitContent const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::installPDFDocument):
(WebKit::UnifiedPDFPlugin::shouldSizeToFitContent const):
(WebKit::UnifiedPDFPlugin::sizeToFitContentsIfNeeded):
(WebKit::UnifiedPDFPlugin::updateLayout):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::PluginView):
(WebKit::PluginView::updateDocumentForPluginSizingBehavior):

Canonical link: https://commits.webkit.org/288932@main
